wire-ios-data-model/Source/MLS/SafeCoreCrypto.swift

```diff
@@ -116,18 +132,36 @@ public class SafeCoreCrypto: SafeCoreCryptoProtocol {
         return result
     }
 
-    // TODO: Remove when async core crypto is supported
-    public func perform<T>(_ block: (WireCoreCrypto.CoreCryptoProtocol) async throws -> T) async rethrows -> T {
-        try await block(coreCrypto)
+    public func perform<T>(_ block: (WireCoreCrypto.CoreCryptoProtocol) throws -> T) async rethrows -> T {
+        var result: T
+        WireLogger.coreCrypto.info("acquiring directory lock")
+        safeContext.acquireDirectoryLock()
+        WireLogger.coreCrypto.info("acquired lock. performing restoreFromDisk()")
+        await restoreFromDisk()
+
+        defer {
+            WireLogger.coreCrypto.info("releasing directory lock")
+            safeContext.releaseDirectoryLock()
+            WireLogger.coreCrypto.info("released lock")
+        }
+
+        do {
+            result = try block(coreCrypto)
+        } catch {
+            WireLogger.coreCrypto.error("failed to perform block on core crypto")
+            throw error
+        }
+
+        return result
     }
 
     public func unsafePerform<T>(_ block: (CoreCryptoProtocol) throws -> T) rethrows -> T {
         return try block(coreCrypto)
     }
 
-    private func restoreFromDisk() {
+    private func restoreFromDisk() async {
```

Comment:
Why not let it throw?

Reply:
Ok, because the surrounding function doesn't throw atm. Let's deal with this later.